### ISSUE TITLE
perf c2s: verify it works on arm64

### ISSUE THREAD

**Issue by adamliyi** [AmpereComputing/ampere-lts-kernel#90]
Quote from patch: https://yhbt.net/lore/all/20201213133850.10070-2-leo.yan@linaro.org/

```
Arm SPE trace data doesn't support HITM, but we still want to explore
"perf c2c" tool to analyze cache false sharing.  If without HITM tag,
the tool cannot give out accurate result for cache false sharing, a
candidate solution is to sort the total load operations and connect with
the threads info, e.g. if multiple threads hit the same cache line for
many times, this can give out the hint that it's likely to cause cache
false sharing issue.
```

It looks latest kernel (5.15-rc1) has merged this patch set, but my test shows perf c2c report print empty result.

Need to verify perf c2c can work on arm64?

```
 perf version
perf version 5.15.rc1.gd0ee23f9d78b

perf c2c record -- ./false_sharing.exe 2
perf c2c report --stdio
=================================================
            Trace Event Information
=================================================
  Total records                     :      62984
  Locked Load/Store Operations      :          0
  Load Operations                   :      62984
  Loads - uncacheable               :          0
  Loads - IO                        :          0
  Loads - Miss                      :       4862
  Loads - no mapping                :          0
  Load Fill Buffer Hit              :          0
  Load L1D hit                      :      58122
  Load L2D hit                      :          0
  Load LLC hit                      :          0
  Load Local HITM                   :          0
  Load Remote HITM                  :          0
  Load Remote HIT                   :          0
  Load Local DRAM                   :          0
  Load Remote DRAM                  :          0
  Load MESI State Exclusive         :          0
  Load MESI State Shared            :          0
  Load LLC Misses                   :          0
  Load access blocked by data       :          0
  Load access blocked by address    :          0
  LLC Misses to Local DRAM          :        nan%
  LLC Misses to Remote DRAM         :        nan%
  LLC Misses to Remote cache (HIT)  :        nan%
  LLC Misses to Remote cache (HITM) :        nan%
  Store Operations                  :          0
  Store - uncacheable               :          0
  Store - no mapping                :          0
  Store L1D Hit                     :          0
  Store L1D Miss                    :          0
  No Page Map Rejects               :        371
  Unable to parse data source       :          0

=================================================
    Global Shared Cache Line Event Information
=================================================
  Total Shared Cache Lines          :          0
  Load HITs on shared lines         :          0
  Fill Buffer Hits on shared lines  :          0
  L1D hits on shared lines          :          0
  Store - no mapping                :          0
  Store L1D Hit                     :          0
  Store L1D Miss                    :          0
  No Page Map Rejects               :        371
  Unable to parse data source       :          0

=================================================
    Global Shared Cache Line Event Information
=================================================
  Total Shared Cache Lines          :          0
  Load HITs on shared lines         :          0
  Fill Buffer Hits on shared lines  :          0
  L1D hits on shared lines          :          0
  L2D hits on shared lines          :          0
  LLC hits on shared lines          :          0
  Locked Access on shared lines     :          0
  Blocked Access on shared lines    :          0
  Store HITs on shared lines        :          0
  Store L1D hits on shared lines    :          0
  Total Merged records              :          0

=================================================
                 c2c details
=================================================
  Events                            : arm_spe_0/ts_enable=1,load_filter=1,store_>
                                    : dummy:u
                                    : memory
  Cachelines sort on                : Total HITMs
  Cacheline data grouping           : offset,iaddr

=================================================
           Shared Data Cache Line Table
=================================================
#
#        ----------- Cacheline ----------      Tot  ------- Load Hitm -------   >
# Index             Address  Node  PA cnt     Hitm    Total  LclHitm  RmtHitm  r>
# .....  ..................  ....  ......  .......  .......  .......  .......  .>
#

=================================================
      Shared Cache Line Distribution Pareto
=================================================
#
(END)
```

**Comment by adamliyi** [comment]
With latest PR: https://github.com/AmpereComputing/ampere-lts-kernel/pull/96


```
perf c2c record -- ./false_sharing.exe 2
perf c2c report -NN -d all --coalesce tid,pid,iaddr,dso --stdio
```

Stores are all '0' - bug?
```
=================================================
           Shared Data Cache Line Table
=================================================
#
#        ----------- Cacheline ----------  Load Hit  Load Hit    Total    Total    Total  ---- Stores ----  ----- Core Load Hit -----  - LLC Load Hit --  - RMT Load Hit --  --- Load Dram ----
# Index             Address  Node  PA cnt       Pct     Total  records    Loads   Stores    L1Hit   L1Miss       FB       L1       L2    LclHit  LclHitm    RmtHit  RmtHitm       Lcl       Rmt
# .....  ..................  ....  ......  ........  ........  .......  .......  .......  .......  .......  .......  .......  .......  ........  .......  ........  .......  ........  ........
#
      0            0x401240   N/A       0    34.61%     32904    32904    32904        0        0        0        0    32904        0         0        0         0        0         0         0
      1      0xffff7bffe980   N/A       0    23.51%     22353    22353    22353        0        0        0        0    22353        0         0        0         0        0         0         0
      2            0x420180   N/A       0    22.74%     21621    21621    21621        0        0        0        0    21621        0         0        0         0        0         0         0
      3      0xffff80cbe980   N/A       0    15.14%     14398    14398    14398        0        0        0        0    14398        0         0        0         0        0         0         0
      4            0x420080   N/A       0     2.89%      2751     2751     2751        0        0        0        0     2751        0         0        0         0        0         0         0
      5            0x420100   N/A       0     0.78%       740      740      740        0        0        0        0      740        0         0        0         0        0         0         0
```

Compared result from:  https://yhbt.net/lore/all/20201213133850.10070-1-leo.yan@linaro.org/

```
The testing result on Arm64 (Hisilicon D06); please note, the Arm SPE
data source patch set has not been merged into the mainline kernel and
a potential issue for store operations is working in progress, so the
final outputting result might have minor differences.

  # perf c2c record -- false_sharing.exe 2
  # perf c2c report -d all --coalesce tid,pid,iaddr,dso --stdio

  [...]

  =================================================
             Shared Data Cache Line Table          
  =================================================
  #
  #        ----------- Cacheline ----------  Load Hit  Load Hit    Total    Total    Total  ---- Stores ----  ----- Core Load Hit -----  - LLC Load Hit --  - RMT Load Hit --  --- Load Dram ----
  # Index             Address  Node  PA cnt       Pct     Total  records    Loads   Stores    L1Hit   L1Miss       FB       L1       L2    LclHit  LclHitm    RmtHit  RmtHitm       Lcl       Rmt
  # .....  ..................  ....  ......  ........  ........  .......  .......  .......  .......  .......  .......  .......  .......  ........  .......  ........  .......  ........  ........
  #
        0      0xaaaab4e8b100   N/A       0    35.04%    100447   104933   100447     4486     4486        0        0    11269        0     89178        0         0        0         0         0
        1      0xaaaab4e8af80   N/A       0    17.29%     49571    49571    49571        0        0        0        0    49571        0         0        0         0        0         0         0
        2      0xaaaab4e8afc0   N/A       0    16.72%     47922    47922    47922        0        0        0        0    47922        0         0        0         0        0         0         0
        3      0xaaaab4e8b080   N/A       0     8.94%     25641    67718    25641    42077    42077        0        0     4397        0     21244        0         0        0         0         0
        4      0xaaaab4e7a480   N/A       0     4.42%     12680    12680    12680        0        0        0        0    12680        0         0        0         0        0         0         0
        5      0xffffa2ffc980   N/A       0     2.62%      7511     7511     7511        0        0        0        0     7511        0         0        0         0        0         0         0
        6      0xffffa3ffe980   N/A       0     2.57%      7374     7374     7374        0        0        0        0     7374        0         0        0         0        0         0         0
        7      0xaaaab4e8b000   N/A       0     2.41%      6907     6907     6907        0        0        0        0     6907        0         0        0         0        0         0         0
        8      0xaaaab4e8b0c0   N/A       0     2.30%      6592     6592     6592        0        0        0        0     2822        0      3770        0         0        0         0         0
        9      0xffffa37fd980   N/A       0     2.24%      6408     6408     6408        0        0        0        0     6408        0         0        0         0        0         0         0
       10      0xffffb8d80980   N/A       0     2.18%      6254     6254     6254        0        0        0        0     6254        0         0        0         0        0         0         0
       11      0xffffb9d82980   N/A       0     1.31%      3763     9706     3763     5943     5943        0        0     3763        0         0        0         0        0         0         0
       12      0xffffb9581980   N/A       0     1.22%      3507    11484     3507     7977     7977        0        0     3507        0         0        0         0        0         0         0
       13      0xffffbad84980   N/A       0     0.33%       932     7766      932     6834     6834        0        0      932        0         0        0         0        0         0         0
       14      0xffffba583980   N/A       0     0.24%       700     6503      700     5803     5803        0        0      700        0         0        0         0        0         0         0
```

**Comment by adamliyi** [comment]
Fix by commits:
edbf18bf0aba perf arm-spe: TESTONLY: fake SPE_ST as hit L1 cache
02febdd0b6f3 perf arm-spe: force SPE_ST record treated as memory sample

perf-c2c display as:

```
# perf c2c record -- ./false_sharing.exe 2
# perf c2c report -NN -d all --coalesce tid,pid,iaddr,dso --stdio

=================================================
            Trace Event Information
=================================================
  Total records                     :     222447
  Locked Load/Store Operations      :          0
  Load Operations                   :     192563
  Loads - uncacheable               :          0
  Loads - IO                        :          0
  Loads - Miss                      :      12107
  Loads - no mapping                :          0
  Load Fill Buffer Hit              :          0
  Load L1D hit                      :     180456
  Load L2D hit                      :          0
  Load LLC hit                      :          0
  Load Local HITM                   :          0
  Load Remote HITM                  :          0
  Load Remote HIT                   :          0
  Load Local DRAM                   :          0
  Load Remote DRAM                  :          0
  Load MESI State Exclusive         :          0
  Load MESI State Shared            :          0
  Load LLC Misses                   :          0
  LLC Misses to Local DRAM          :        nan%
  LLC Misses to Remote DRAM         :        nan%
  LLC Misses to Remote cache (HIT)  :        nan%
  LLC Misses to Remote cache (HITM) :        nan%
  Store Operations                  :      29884  <------------------- Store is reported
  Store - uncacheable               :          0
  Store - no mapping                :          0
  Store L1D Hit                     :      29884  <--------------------- Fake store as Hit L1 cache
  Store L1D Miss                    :          0
  No Page Map Rejects               :        834
  Unable to parse data source       :          0
```

```
=================================================
      Shared Cache Line Distribution Pareto
=================================================
#
#        --  Load Refs --  -- Store Refs --  --------- Data address ---------                                                        ---------- cycles ----------    Total       cpu                                  Sha>
#   Num      Hit     Miss   L1 Hit  L1 Miss              Offset  Node  PA cnt      Pid                      Tid        Code address  rmt hitm  lcl hitm      load  records       cnt               Symbol             Obj>
# .....  .......  .......  .......  .......  ..................  ....  ......

 -------------------------------------------------------------
      3    30302        0     7220        0            0x420180
  -------------------------------------------------------------
          64.32%    0.00%    0.00%    0.00%                 0x0   N/A       0  1099889  1099889:false_sharing.e            0x400bd0         0         0         0    19490         2  [.] read_write_func  false_sharing.>
          15.99%    0.00%    0.00%    0.00%                 0x0   N/A       0  1099889  1099889:false_sharing.e            0x400be8         0         0         0     4844         2  [.] read_write_func  false_sharing.>
           0.00%    0.00%   73.91%    0.00%                 0x0   N/A       0  1099889  1099889:false_sharing.e            0x400bd4         0         0         0     5336         2  [.] read_write_func  false_sharing.>
           0.00%    0.00%   26.09%    0.00%                 0x0   N/A       0  1099889  1099889:false_sharing.e            0x400bf8         0         0         0     1884         2  [.] read_write_func  false_sharing.>
          19.70%    0.00%    0.00%    0.00%                0x20   N/A       0  1099889  1099889:false_sharing.e            0x400c64         0         0         0     5968         2  [.] read_write_func  false_sharing.
```
